core/src/main/scala/org/apache/spark/sql/internal/SnappySessionState.scala

```diff
@@ -102,14 +102,18 @@ class SnappySessionState(val snappySession: SnappySession)
 
     override lazy val batches: Seq[Batch] = analyzer.batches.map {
       case batch if batch.name.equalsIgnoreCase("Resolution") =>
-        Batch(batch.name, getStrategy(batch.strategy), batch.rules.filter(_ match {
+        val rules = batch.rules.filter(_ match {
           case PromoteStrings => if (sqlParser.sqlParser.questionMarkCounter > 0) {
             false
           } else {
             true
           }
           case _ => true
-        }): _*)
+        }).flatMap(_ match {
+          case PromoteStrings => Seq(StringPromotionCheckForUpdate, PromoteStrings)
+          case a => Seq(a)
+        })
```

Comment:
Can't the above filter rule be combined in flatMap instead of traversing twice?

```scala
        }).flatMap(_ match {
          case PromoteStrings => if (sqlParser.sqlParser.questionMarkCounter > 0) Nil
                                 else StringPromotionCheckForUpdate :: PromoteStrings :: Nil
          ...
```
Its slightly more efficient to use "::" for small sequences (a::Nil) and Nil avoids object creation and always better than Seq.empty